File: android/app/src/main/kotlin/org/fgsec/halogen/features/latest/LatestView.kt
package org.fgsec.halogen.features.latest

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyListState
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshotFlow
import androidx.compose.ui.Modifier
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.compose.LocalLifecycleOwner
import kotlinx.coroutines.flow.drop
import kotlinx.coroutines.launch
import org.fgsec.halogen.components.EmptyStateView
import org.fgsec.halogen.components.HalogenNavbar
import org.fgsec.halogen.components.ListControlsBar
import org.fgsec.halogen.components.ListQuery
import org.fgsec.halogen.components.LoadErrorView
import org.fgsec.halogen.components.LoadMoreRow
import org.fgsec.halogen.core.HalogenCore
import org.fgsec.halogen.core.SwipePage
import org.fgsec.halogen.features.podcasts.BulkActionBar
import org.fgsec.halogen.features.podcasts.SelectableEpisodeRow

/// The Latest tab: newest episodes across every subscription, faceted by the
/// filter bar, local-first, with infinite scroll, restored scroll position,
/// configurable swipes, quick-action row menus, and bulk multiselect.
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun LatestView(model: LatestModel, core: HalogenCore) {
    val scope = rememberCoroutineScope()
    var selection by remember { mutableStateOf(setOf<Int>()) }
    var selectedOnly by remember { mutableStateOf(false) }
    /// Row selection is an EDIT-MODE tool only: rows keep their normal taps
    /// until Edit is active, and leaving Edit clears the set (closes the
    /// bulk bar).
    var isEditing by remember { mutableStateOf(false) }
    var refreshing by remember { mutableStateOf(false) }
    val listState: LazyListState =
        rememberSaveable(saver = LazyListState.Saver) { LazyListState() }

    // task(id: query): first appearance and every query change reload.
    LaunchedEffect(model.query) { model.load() }

    // OnDevice facet: the list is a one-shot copy of the device set —
    // recompute when membership changes (a download completing or a
    // removal), or the facet shows stale rows until a manual refresh.
    LaunchedEffect(Unit) {
        snapshotFlow { core.models?.device?.onDevice?.map { it.id } ?: emptyList() }
            .drop(1)
            .collect {
                if (model.query.filters.contains(EpisodeFilter.OnDevice)) model.refresh()
            }
    }

    // Backgrounding (and tab disappear) persist the scroll anchor — it
    // survives relaunch, not just tab switches.
    val lifecycleOwner = LocalLifecycleOwner.current
    DisposableEffect(lifecycleOwner) {
        val observer = LifecycleEventObserver { _, event ->
            if (event == Lifecycle.Event.ON_PAUSE) model.persistScrollAnchor()
        }
        lifecycleOwner.lifecycle.addObserver(observer)
        onDispose {
            lifecycleOwner.lifecycle.removeObserver(observer)
            model.persistScrollAnchor()
        }
    }

    /// The "Selected" review chip restricts rows to the ticked set.
    val displayed =
        if (selectedOnly) model.episodes.filter { it.id in selection } else model.episodes

    // One-shot scroll restore: consume the model's pending target.
    val pending = model.pendingScrollTo
    LaunchedEffect(pending) {
        if (pending != null) {
            val idx = displayed.indexOfFirst { it.id == pending }
            if (idx >= 0) listState.scrollToItem(idx)
            model.pendingScrollTo = null
        }
    }

    Column(Modifier.fillMaxSize()) {
        HalogenNavbar(core = core, leading = {
            TextButton(onClick = {
                isEditing = !isEditing
                // Leaving Edit closes multi-select for real (the bulk bar is
                // keyed on a non-empty selection).
                if (!isEditing) selection = emptySet()
            }) {
                Text(if (isEditing) "Done" else "Edit")
            }
        })
        ListControlsBar(
            query = model.query,
            onQueryChange = { model.query = it },
            allowsOnDevice = !core.isEmbeddedAccount,
        )
        Box(Modifier.fillMaxSize().weight(1f)) {
            val error = model.error
            when {
                error != null -> LoadErrorView(
                    title = "Couldn't load latest",
                    message = error,
                    retry = { model.refresh() },
                )
                model.loaded && model.episodes.isEmpty() -> EmptyStateView(
                    systemImage = "clock",
                    title = "Nothing here",
                    description = if (model.query == ListQuery())
                        "New episodes land here after each feed poll."
                    else
                        "Nothing matches the current search/filter.",
                )
                else -> PullToRefreshBox(
                    isRefreshing = refreshing,
                    onRefresh = {
                        scope.launch {
                            refreshing = true
                            try {
                                model.refresh()
                            } finally {
                                refreshing = false
                            }
                        }
                    },
                ) {
                    LazyColumn(state = listState, modifier = Modifier.fillMaxSize()) {
                        items(count = displayed.size, key = { displayed[it].id }) { index ->
                            val episode = displayed[index]
                            DisposableEffect(episode.id) {
                                model.rowAppeared(episode)
                                onDispose { model.rowDisappeared(episode) }
                            }
                            SelectableEpisodeRow(
                                episode = episode,
                                subtitle = episode.podcast?.title,
                                page = SwipePage.Latest,
                                isEditing = isEditing,
                                selected = episode.id in selection,
                                onToggleSelect = {
                                    selection = if (episode.id in selection)
                                        selection - episode.id
                                    else selection + episode.id
                                },
                                core = core,
                            )
                        }
                        if (model.hasMore && model.episodes.isNotEmpty() && !selectedOnly) {
                            item {
                                LoadMoreRow(failed = model.loadMoreFailed) { model.loadMore() }
                            }
                        }
                    }
                }
            }
        }
        if (selection.isNotEmpty()) {
            BulkActionBar(
                selection = selection,
                onSelectionChange = { selection = it },
                selectedOnly = selectedOnly,
                onSelectedOnlyChange = { selectedOnly = it },
                episodes = model.episodes,
                core = core,
            )
        }
    }
}

File: ios/Halogen/Features/Latest/LatestView.swift
import SwiftUI

/// The Latest tab: newest episodes across every subscription, faceted by the
/// filter bar, local-first, with infinite scroll, restored scroll position,
/// configurable swipes, quick-action row menus, and bulk multiselect.
struct LatestView: View {
    @Bindable var model: LatestModel
    let core: HalogenCore

    @Environment(\.scenePhase) private var scenePhase
    @State private var selection = Set<Int32>()
    @State private var selectedOnly = false
    /// Row selection is an EDIT-MODE tool only: iOS 17 lists tap-select
    /// outside edit mode too, which hijacked plain row taps into an
    /// unclosable multi-select. Rows disable selection until Edit is active,
    /// and leaving Edit clears the set (closes the bulk bar).
    @Environment(\.editMode) private var editMode

    private var isEditing: Bool { editMode?.wrappedValue.isEditing == true }

    var body: some View {
        content
            // safeAreaInset (not a plain VStack) so the bar keeps its own
            // layout slot under iOS 26's glass navbar instead of being
            // overlapped by its scroll-edge surface.
            .safeAreaInset(edge: .top, spacing: 0) {
                ListControlsBar(query: $model.query, allowsOnDevice: !core.isEmbeddedAccount)
            }
            .safeAreaInset(edge: .bottom, spacing: 0) {
                if !selection.isEmpty {
                    BulkActionBar(
                        selection: $selection, selectedOnly: $selectedOnly,
                        episodes: model.episodes, core: core)
                }
            }
            .halogenNavbar(core: core)
            .toolbar {
                ToolbarItem(placement: .topBarLeading) {
                    EditButton()
                }
            }
            .task(id: model.query) { await model.load() }
            .refreshable { await model.refresh() }
            // Leaving Edit closes multi-select for real (the bulk bar is
            // keyed on a non-empty selection).
            .onChange(of: isEditing) { _, editing in
                if !editing { selection.removeAll() }
            }
            // OnDevice facet: the list is a one-shot copy of the device set —
            // recompute when membership changes (a download completing or a
            // removal), or the facet shows stale rows until a manual refresh.
            .onChange(of: core.models?.device.onDevice.map(\.id) ?? []) {
                guard model.query.filters.contains(.onDevice) else { return }
                Task { await model.refresh() }
            }
            .onChange(of: scenePhase) { _, phase in
                if phase != .active { model.persistScrollAnchor() }
            }
            .onDisappear { model.persistScrollAnchor() }
    }

    @ViewBuilder
    private var content: some View {
        if let error = model.error {
            ContentUnavailableView {
                Label("Couldn't load latest", systemImage: "wifi.exclamationmark")
            } description: {
                Text(error).font(.footnote.monospaced())
            }
        } else if model.loaded && model.episodes.isEmpty {
            ContentUnavailableView(
                "Nothing here",
                systemImage: "clock",
                description: Text(emptyHint)
            )
        } else {
            episodeList
        }
    }

    /// The "Selected" review chip restricts rows to the ticked set.
    private var displayedEpisodes: [EpisodeData] {
        selectedOnly ? model.episodes.filter { selection.contains($0.id) } : model.episodes
    }

    private var episodeList: some View {
        ScrollViewReader { proxy in
            List(selection: $selection) {
                ForEach(displayedEpisodes, id: \.id) { episode in
                    EpisodeRowLink(
                        episode: episode,
                        artURL: core.episodeArtURL(episode),
                        subtitle: episode.podcast?.title,
                        context: .browse,
                        core: core
                    )
                    .tag(episode.id)
                    .selectionDisabled(!isEditing)
                    .id(episode.id)
                    .configuredSwipes(.latest, episode: episode, core: core)
                    .onAppear { model.rowAppeared(episode) }
                    .onDisappear { model.rowDisappeared(episode) }
                }
                if model.hasMore && !model.episodes.isEmpty && !selectedOnly {
                    LoadMoreRow(failed: model.loadMoreFailed) { await model.loadMore() }
                }
            }
            .listStyle(.plain)
            .onChange(of: model.pendingScrollTo) { _, target in
                if let target {
                    proxy.scrollTo(target, anchor: .top)
                    model.pendingScrollTo = nil
                }
            }
        }
    }

    private var emptyHint: String {
        model.query == ListQuery()
            ? "New episodes land here after each feed poll."
            : "Nothing matches the current search/filter."
    }
}
